File: platform/macos/Desktop/NativeRemoteResizeCoordinator.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Combine
import Foundation

// One coordinator per session. Geometry input has a view identity; removing an
// older view must not invalidate its replacement. Only one source is authoritative.
struct NativeResizeViewport: Equatable {
  let width: Double, height: Double, scale: Double
  let unscaled: Bool, devicePixels: Bool, available: Bool
}
@MainActor public final class NativeRemoteResizeCoordinator: ObservableObject {
  @Published public private(set) var message: String?
  private weak var session: NativeSession?
  private var observations = Set<AnyCancellable>()
  private var owner: UUID?
  private var viewport: NativeResizeViewport?
  // A fullscreen canvas reserves geometry ownership even while unavailable.
  // Ordinary source-window updates are remembered, never used during the lease.
  private var canvasOwner: UUID?
  private var canvasGeometry: Geometry?
  private enum Geometry: Equatable {
    case window(NativeResizeViewport)
    case canvas(NativeDisplayLayout, unscaled: Bool, available: Bool)
    var available: Bool {
      switch self { case .window(let value): value.available; case .canvas(_,_,let value): value }
    }
  }
  private var geometry: Geometry? { canvasOwner == nil ? viewport.map(Geometry.window) : canvasGeometry }
  private var geometryOwner: UUID? { canvasOwner ?? owner }
  private var delay: Task<Void,Never>?, operation: Task<Void,Never>?, wake: Task<Void,Never>?
  private var generation: UInt64 = 0
  private var initialPolicy: NativeRemoteResizePolicy = .builtIn
  private var initialAttempted = false
  private var initialViewport: Geometry?
  private var manualViewport: Geometry?
  private var lastAttempt: Target?
  private var stopped = false
  private struct Target: Equatable {
    let width: UInt32, height: UInt32, generation: UInt64
    let revision: UUID
    let initial: Bool
    let owner: UUID
    let canvas: NativeDisplayLayout?
  }
  init(session: NativeSession) {
    self.session = session
    session.$snapshot.removeDuplicates { a,b in
      a.generation == b.generation && a.state == b.state && a.supportsResize == b.supportsResize && a.resizePending == b.resizePending
    }.sink { [weak self] _ in self?.schedule() }.store(in:&observations)
    session.$isViewOnly.removeDuplicates().sink { [weak self] _ in self?.schedule() }.store(in:&observations)
    session.$resizePolicy.removeDuplicates().sink { [weak self] value in
      if !value.enabled { self?.operation?.cancel() }; self?.schedule()
    }.store(in:&observations)
    session.$isClosing.sink { [weak self] value in if value { self?.stop() } }.store(in:&observations)
  }
  deinit { delay?.cancel(); operation?.cancel(); wake?.cancel() }
  func beginAttempt(generation: UInt64, policy: NativeRemoteResizePolicy) {
    self.generation = generation; initialPolicy = policy
    initialAttempted = false; initialViewport = nil; manualViewport = nil; lastAttempt = nil; message = nil
    schedule()
  }
  func attach(owner: UUID) {
    guard !stopped, canvasOwner == nil, self.owner != owner else { return }
    self.owner = owner; viewport = nil; lastAttempt = nil; schedule()
  }
  func manualRequest() {
    manualViewport = geometry; lastAttempt = nil; delay?.cancel(); delay = nil
  }
  func update(owner: UUID, viewport: NativeResizeViewport) {
    guard !stopped else { return }
    guard self.owner == owner || (self.owner == nil && canvasOwner == nil && viewport.available) else { return }
    // Attachment explicitly chooses the authoritative geometry source.
    guard self.owner != owner || self.viewport != viewport else { return }
    self.owner = owner; self.viewport = viewport; schedule()
  }
  func detach(owner: UUID) {
    guard self.owner == owner else { return }
    self.owner = nil; viewport = nil; delay?.cancel(); delay = nil; operation?.cancel(); lastAttempt = nil
    // Sent requests still drain; no later viewport work is scheduled while detached.
  }
  func beginCanvas(owner: UUID) {
    guard !stopped, canvasOwner != owner else { return }
    canvasOwner = owner; canvasGeometry = nil; lastAttempt = nil
    delay?.cancel(); delay = nil; operation?.cancel(); schedule()
  }
  func updateCanvas(owner: UUID, layout: NativeDisplayLayout, unscaled: Bool, available: Bool) {
    guard !stopped, canvasOwner == owner else { return }
    let value = Geometry.canvas(layout,unscaled:unscaled,available:available)
    guard canvasGeometry != value else { return }
    canvasGeometry = value
    if !available { delay?.cancel(); delay = nil; operation?.cancel() }
    schedule()
  }
  func endCanvas(owner: UUID) {
    guard canvasOwner == owner else { return }
    canvasOwner = nil; canvasGeometry = nil; lastAttempt = nil
    delay?.cancel(); delay = nil; operation?.cancel(); schedule()
  }
  private func schedule() {
    guard !stopped, wake == nil else { return }
    wake = Task { @MainActor [weak self] in
      guard let self else { return }; self.wake = nil; self.reconcile()
    }
  }
  private func target() -> Target? {
    guard let session else { return nil }
    if generation != session.generation {
      generation = session.generation; initialPolicy = session.resizePolicy
      initialAttempted = false; initialViewport = nil; manualViewport = nil; lastAttempt = nil; message = nil
    }
    guard !stopped, !session.isClosing, session.snapshot.state == .connected,
          session.snapshot.supportsResize, !session.snapshot.resizePending, !session.isViewOnly,
          session.resizePolicy.enabled, let geometry, geometry.available, let owner = geometryOwner else { return nil }
    if let manualViewport {
      guard geometry != manualViewport else { return nil }
      self.manualViewport = nil // A later return to this geometry is a new resize.
    }
    if let initialViewport, geometry != initialViewport { self.initialViewport = nil }
    if !initialAttempted, let w = initialPolicy.initialWidth, let h = initialPolicy.initialHeight {
      return Target(width:w,height:h,generation:generation,revision:session.resizePolicyRevision,initial:true,owner:owner,canvas:nil)
    }
    guard geometry != initialViewport else { return nil }
    if case .canvas(let layout,let unscaled,_) = geometry {
      guard unscaled else { return nil }
      return Target(width:layout.width,height:layout.height,generation:generation,revision:session.resizePolicyRevision,
        initial:false,owner:owner,canvas:layout)
    }
    guard case .window(let viewport) = geometry, viewport.unscaled, viewport.width.isFinite, viewport.height.isFinite,
          viewport.scale.isFinite, viewport.scale > 0 else { return nil }
    let scale = viewport.devicePixels ? viewport.scale : 1
    let width = floor(viewport.width * scale), height = floor(viewport.height * scale)
    guard width >= 1, height >= 1, width <= 65535, height <= 65535 else { return nil }
    return Target(width:UInt32(width),height:UInt32(height),generation:generation,revision:session.resizePolicyRevision,initial:false,owner:owner,canvas:nil)
  }
  private func reconcile() {
    delay?.cancel(); delay = nil
    guard operation == nil, let target = target(), target != lastAttempt else { return }
    delay = Task { @MainActor [weak self] in
      do { try await Task.sleep(for:.milliseconds(100)) } catch { return }
      guard let self, !Task.isCancelled, self.target() == target else { return }
      self.delay = nil; self.send(target)
    }
  }
  private func send(_ target: Target) {
    guard let session, operation == nil, self.target() == target else { return }
    do {
      let current = try session.desktopLayout()
      guard let first = current.layout.screens.first else { return }
      let requested: NativeRemoteLayout
      if let canvas = target.canvas { requested = try canvas.remoteLayout(matching:current.layout) }
      else { requested = try NativeRemoteLayout(width:target.width,height:target.height,
        screens:[.init(id:first.id,x:0,y:0,width:target.width,height:target.height,flags:first.flags)]) }
      lastAttempt = target; if target.initial { initialAttempted = true; initialViewport = geometry }
      if requested == current.layout { schedule(); return }
      message = nil
      operation = Task { @MainActor [weak self] in
        do { _ = try await session.requestAutomaticDesktopLayout(requested,expectedGeneration:target.generation) }
        catch {
          guard let self else { return }
          if !self.stopped && session.generation == target.generation && self.geometryOwner == target.owner {
            if let error = error as? NativeError, error.status == .busy {
              self.lastAttempt = nil; if target.initial { self.initialAttempted = false; self.initialViewport = nil }
            } else if !(error is CancellationError) {
              self.message = String(localized:"desktop.resize.the.automatic.desktop.resize.did.not.complete.you.can.request.a.size", defaultValue:"The automatic desktop resize did not complete. You can request a size with Resize Remote Desktop.")
            }
          }
        }
        guard let self else { return }; self.operation = nil; self.schedule()
      }
    } catch { message = String(localized:"desktop.resize.the.remote.desktop.layout.is.unavailable.for.automatic.resizing", defaultValue:"The remote desktop layout is unavailable for automatic resizing.") }
  }
  public func stop() {
    stopped = true; observations.removeAll(); delay?.cancel(); delay = nil; wake?.cancel(); wake = nil
    operation?.cancel(); viewport = nil; owner = nil; canvasGeometry = nil; canvasOwner = nil
  }
  func close() async { stop(); await operation?.value }
}

File: platform/windows/TidyVNC.Native/Desktop/NativeRemoteResizeCoordinator.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using System.ComponentModel;
using CommunityToolkit.Mvvm.ComponentModel;
using TidyVNC.Native.Platform;

namespace TidyVNC.Native.Desktop;

/// <summary>A desktop view's size for automatic resizing: effective pixels, its scale and the scaling it shows.</summary>
public readonly record struct NativeResizeViewport(double Width, double Height, double Scale, bool Unscaled, bool DevicePixels, bool Available);

/// <summary>
/// Automatic remote resize for one connection (macOS
/// NativeRemoteResizeCoordinator; REMOTE-RESIZE.md, PARITY D06-D07). One
/// geometry source is authoritative: the window's desktop view, or, while
/// full screen spans displays, the canvas of its surfaces (window updates are
/// remembered but not used meanwhile). Requests wait 100 ms for the geometry
/// to settle, one runs at a time, and a size is never requested twice for
/// the same geometry. The initial size is requested once per connection
/// when resizing is enabled; after a manual request, automatic resizing
/// waits for the geometry to change. UI thread only.
/// </summary>
public sealed partial class NativeRemoteResizeCoordinator : ObservableObject, IDisposable
{
    private abstract record Geometry(bool Available);
    private sealed record WindowGeometry(NativeResizeViewport Viewport) : Geometry(Viewport.Available);
    private sealed record CanvasGeometry(NativeDisplayLayout Layout, bool Unscaled, bool IsAvailable) : Geometry(IsAvailable);
    private sealed record Target(uint Width, uint Height, ulong Generation, Guid Revision, bool Initial, Guid Owner, NativeDisplayLayout? Canvas);

    private NativeSession? session;
    private Guid? owner, canvasOwner;
    private NativeResizeViewport? viewport;
    private CanvasGeometry? canvasGeometry;
    private CancellationTokenSource? delay;
    private Task? operation;
    private bool wakeQueued, stopped, initialAttempted;
    private ulong generation;
    private NativeRemoteResizePolicy initialPolicy = NativeRemoteResizePolicy.BuiltIn;
    private Geometry? initialViewport, manualViewport;
    private Target? lastAttempt;

    /// <summary>Why the last automatic resize did not happen, or null.</summary>
    [ObservableProperty] public partial NativeText? Message { get; private set; }

    /// <summary>How long the geometry must stay unchanged before a request (tests shorten it).</summary>
    public TimeSpan Settle { get; init; } = TimeSpan.FromMilliseconds(100);

    private Geometry? Current => canvasOwner is null ? viewport is { } value ? new WindowGeometry(value) : null : canvasGeometry;
    private Guid? GeometryOwner => canvasOwner ?? owner;

    public void Bind(NativeSession value)
    {
        if (stopped || ReferenceEquals(session, value)) return;
        Unbind();
        session = value;
        generation = 0;
        value.PropertyChanged += SessionChanged;
        value.ManualResizeRequested += ManualRequest;
        Schedule();
    }

    private void Unbind()
    {
        if (session is null) return;
        session.PropertyChanged -= SessionChanged;
        session.ManualResizeRequested -= ManualRequest;
        session = null;
    }

    private void SessionChanged(object? sender, PropertyChangedEventArgs e)
    {
        switch (e.PropertyName)
        {
            case nameof(NativeSession.IsClosing) when session?.IsClosing == true:
                Stop();
                return;
            case nameof(NativeSession.ResizePolicy) when session?.ResizePolicy.Enabled == false:
                CancelDelay();
                Schedule();
                return;
            case nameof(NativeSession.Snapshot) or nameof(NativeSession.IsViewOnly) or nameof(NativeSession.ResizePolicy):
                Schedule();
                return;
        }
    }

    /// <summary>A manual request: keep that size until the geometry changes.</summary>
    private void ManualRequest()
    {
        manualViewport = Current;
        lastAttempt = null;
        CancelDelay();
    }

    /// <summary>The window's desktop view reports its size; the first available view becomes the source.</summary>
    public void Update(Guid source, NativeResizeViewport value)
    {
        if (stopped) return;
        if (owner != source && !(owner is null && canvasOwner is null && value.Available)) return;
        if (owner == source && viewport == value) return;
        owner = source;
        viewport = value;
        Schedule();
    }

    public void Detach(Guid source)
    {
        if (owner != source) return;
        owner = null; viewport = null; lastAttempt = null;
        CancelDelay();
    }

    /// <summary>Full screen across displays takes the geometry over, even while not yet shown.</summary>
    public void BeginCanvas(Guid source)
    {
        if (stopped || canvasOwner == source) return;
        canvasOwner = source; canvasGeometry = null; lastAttempt = null;
        CancelDelay();
        Schedule();
    }

    public void UpdateCanvas(Guid source, NativeDisplayLayout layout, bool unscaled, bool available)
    {
        if (stopped || canvasOwner != source) return;
        var value = new CanvasGeometry(layout, unscaled, available);
        if (canvasGeometry == value) return;
        canvasGeometry = value;
        if (!available) CancelDelay();
        Schedule();
    }

    public void EndCanvas(Guid source)
    {
        if (canvasOwner != source) return;
        canvasOwner = null; canvasGeometry = null; lastAttempt = null;
        CancelDelay();
        Schedule();
    }

    private void CancelDelay()
    {
        delay?.Cancel();
        delay?.Dispose();
        delay = null;
    }

    private void Schedule()
    {
        if (stopped || wakeQueued) return;
        wakeQueued = true;
        _ = Wake();
    }

    private async Task Wake()
    {
        await Task.Yield();
        wakeQueued = false;
        if (!stopped) Reconcile();
    }

    private Target? Compute()
    {
        if (session is not { } s) return null;
        if (generation != s.Generation)
        {
            generation = s.Generation; initialPolicy = s.ResizePolicy;
            initialAttempted = false; initialViewport = null; manualViewport = null; lastAttempt = null; Message = null;
        }
        if (stopped || s.IsClosing || s.Snapshot is not { State: NativeSessionState.Connected, SupportsResize: true, ResizePending: false } ||
            s.IsViewOnly || !s.ResizePolicy.Enabled || Current is not { Available: true } geometry || GeometryOwner is not { } source) return null;
        if (manualViewport is not null)
        {
            if (geometry == manualViewport) return null;
            manualViewport = null; // A later return to that geometry is a new resize.
        }
        if (initialViewport is not null && geometry != initialViewport) initialViewport = null;
        if (!initialAttempted && initialPolicy is { InitialWidth: { } w, InitialHeight: { } h })
            return new Target(w, h, generation, s.ResizePolicyRevision, true, source, null);
        if (geometry == initialViewport) return null;
        if (geometry is CanvasGeometry canvas)
            return canvas.Unscaled ? new Target(canvas.Layout.Width, canvas.Layout.Height, generation, s.ResizePolicyRevision, false, source, canvas.Layout) : null;
        if (geometry is not WindowGeometry { Viewport: { Unscaled: true } view } || !double.IsFinite(view.Width) || !double.IsFinite(view.Height) ||
            !double.IsFinite(view.Scale) || view.Scale <= 0) return null;
        var scale = view.DevicePixels ? view.Scale : 1;
        double width = Math.Floor(view.Width * scale), height = Math.Floor(view.Height * scale);
        if (width is < 1 or > 65535 || height is < 1 or > 65535) return null;
        return new Target((uint)width, (uint)height, generation, s.ResizePolicyRevision, false, source, null);
    }

    private void Reconcile()
    {
        CancelDelay();
        if (operation is not null || Compute() is not { } target || target == lastAttempt) return;
        delay = new CancellationTokenSource();
        _ = Settled(target, delay.Token);
    }

    private async Task Settled(Target target, CancellationToken token)
    {
        try { await Task.Delay(Settle, token); }
        catch (OperationCanceledException) { return; }
        if (stopped || token.IsCancellationRequested || Compute() != target) return;
        CancelDelay();
        Send(target);
    }

    private void Send(Target target)
    {
        if (session is not { } s || operation is not null || Compute() != target) return;
        NativeRemoteLayout requested;
        try
        {
            var current = s.DesktopLayout();
            if (current.Layout.Screens.Count == 0) return;
            var first = current.Layout.Screens[0];
            requested = target.Canvas is { } canvas ? canvas.RemoteLayout(current.Layout)
                : new NativeRemoteLayout(target.Width, target.Height, [new NativeRemoteScreen(first.Id, 0, 0, target.Width, target.Height, first.Flags)]);
            lastAttempt = target;
            if (target.Initial) { initialAttempted = true; initialViewport = Current; }
            if (requested.Equals(current.Layout)) { Schedule(); return; }
        }
        catch (NativeError)
        {
            Message = new NativeText("desktop.resize.the.remote.desktop.layout.is.unavailable.for.automatic.resizing");
            return;
        }
        if (Current is WindowGeometry { Viewport: var measured })
            NativeProcessLogging.Viewport(measured.Width, measured.Height, measured.Scale);
        Message = null;
        operation = Run(s, requested, target);
    }

    private async Task Run(NativeSession s, NativeRemoteLayout requested, Target target)
    {
        try { await s.RequestAutomaticDesktopLayoutAsync(requested, target.Generation); }
        catch (Exception error) when (error is NativeError or NativeCommandFailure or OperationCanceledException)
        {
            if (!stopped && s.Generation == target.Generation && GeometryOwner == target.Owner)
            {
                if (error is NativeError { Status: NativeStatus.Busy })
                {
                    lastAttempt = null;
                    if (target.Initial) { initialAttempted = false; initialViewport = null; }
                }
                else if (error is not OperationCanceledException)
                    Message = new NativeText("desktop.resize.the.automatic.desktop.resize.did.not.complete.you.can.request.a.size");
            }
        }
        operation = null;
        Schedule();
    }

    public void Stop()
    {
        if (stopped) return;
        stopped = true;
        CancelDelay();
        Unbind();
        owner = null; viewport = null; canvasOwner = null; canvasGeometry = null;
    }

    public void Dispose() => Stop();

    /// <summary>Stops and waits for a request already sent (close ordering).</summary>
    public async Task CloseAsync()
    {
        Stop();
        if (operation is { } running) await running;
    }
}
